ui/components/app/currency-input/currency-input.js

```diff
@@ -41,84 +45,93 @@ export default function CurrencyInput({
 
-  const getDecimalValue = () => {
-    const decimalValueString = shouldUseFiat
-      ? getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: secondaryCurrency,
-          conversionRate,
-          numberOfDecimals: 2,
-        })
-      : getValueFromWeiHex({
-          value: hexValue,
-          toCurrency: EtherDenomination.ETH,
-          numberOfDecimals: 8,
-        });
-
-    return Number(decimalValueString) || 0;
-  };
+  const tokenToFiatConversionRate = useTokenExchangeRate(asset?.address);
 
-  const initialDecimalValue = hexValue ? getDecimalValue() : 0;
+  const processNewDecimalValue = useProcessNewDecimalValue(
+    assetDecimals,
+    !isTokenPrimary,
+    tokenToFiatConversionRate,
+  );
 
   const swap = async () => {
     await onPreferenceToggle();
-    setSwapped(!isSwapped);
     setShouldDisplayFiat(!shouldDisplayFiat);
   };
 
+  // if the conversion rate is undefined, do not allow a fiat input
+  useEffect(() => {
+    if (isTokenPrimary) {
+      return;
+    }
+
+    if (!tokenToFiatConversionRate) {
+      onPreferenceToggle();
+      setShouldDisplayFiat(false);
+    }
+  }, [tokenToFiatConversionRate, isTokenPrimary, onPreferenceToggle]);
+
   const handleChange = (newDecimalValue) => {
-    const hexValueNew = shouldUseFiat
-      ? getWeiHexFromDecimalValue({
-          value: newDecimalValue,
-          fromCurrency: secondaryCurrency,
-          conversionRate,
-          invertConversionRate: true,
-        })
-      : getWeiHexFromDecimalValue({
-          value: newDecimalValue,
-          fromCurrency: EtherDenomination.ETH,
-          fromDenomination: EtherDenomination.ETH,
-          conversionRate,
-        });
-
-    setNewHexValue(hexValueNew);
-    onChange(hexValueNew);
-    setSwapped(!isSwapped);
+    const { newTokenDecimalValue, newFiatDecimalValue } =
+      processNewDecimalValue(newDecimalValue);
+    setTokenDecimalValue(newTokenDecimalValue);
+    setFiatDecimalValue(newFiatDecimalValue);
+
+    onChange(
+      new Numeric(newTokenDecimalValue, 10)
+        .times(Math.pow(10, assetDecimals), 10)
+        .toPrefixedHexString(),
+    );
   };
 
   useEffect(() => {
-    setNewHexValue(hexValue);
-  }, [hexValue]);
-
-  useEffect(() => {
-    if (featureSecondary) {
-      handleChange(initialDecimalValue);
+    // do not override the input when it is using fiat, since it is imprecise
+    if (!isTokenPrimary) {
+      return;
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [featureSecondary, initialDecimalValue]);
+
+    const decimalizedHexValue = new Numeric(hexValue, 16)
+      .toBase(10)
+      .shiftedBy(assetDecimals)
+      .toString();
+
+    const { newTokenDecimalValue, newFiatDecimalValue } =
+      processNewDecimalValue(decimalizedHexValue);
+
+    setTokenDecimalValue(newTokenDecimalValue);
+    setFiatDecimalValue(newFiatDecimalValue);
+  }, [hexValue, assetDecimals, processNewDecimalValue, isTokenPrimary]);
```

Comment:
What's the difference between this and `handleChange`? Wouldn't this cause the new decimal values to be calculated and set twice on change?

Reply:
This component is controlled, so this ensures consistency with the passed `hexValue`

Reply:
I meant consistency with the passed prop since it’s controlled

Reply:
Based on my understanding, handleChange updates `hexValue`, so the useEffect would be re-updating local state with the same values on change. But this covers scenarios in which non-user interactions affect the display values, for example: when exchangeRates update in the background, the processValues callback is recreated, then this useEffect runs, which updates the decimalValue states

However, it seems like the complexity can be reduced by memoizing the decimalValues based on hexValue, asset, shouldDisplayFiat, exchangeRates, etc (useMemo instead of useState + useEffect + useCallback); handleChange can be simplified with this approach too, since it won't need to directly maintain the decimalValue states